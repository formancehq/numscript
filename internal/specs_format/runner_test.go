package specs_format_test

import (
	"bytes"
	"testing"

	"github.com/formancehq/numscript/internal/specs_format"
	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/stretchr/testify/require"
)

func TestShowDiff(t *testing.T) {
	var buf bytes.Buffer
	specs_format.ShowDiff(
		&buf,
		map[string]any{
			"common": "ok",
			"x":      42,
		},
		map[string]any{
			"common": "ok",
			"x":      100,
		},
	)
	snaps.MatchSnapshot(t, buf.String())
}

func TestSingleTest(t *testing.T) {
	var out bytes.Buffer

	script := `
		send [USD/2 100] (
			source = @world
			destination = @dest
		)
	`

	specs := `
		{
			"testCases": [
				{
					"it": "tfailing",
					"expect.postings": [{
						"source": "wrong-source",
						"destination": "dest",
						"asset": "USD/2",
						"amount": 100
					}]
				},
				{
					"it": "tpassing",
					"expect.postings": [{
						"source": "world",
						"destination": "dest",
						"asset": "USD/2",
						"amount": 100
					}]
				}
			] 
		}
	`

	success := specs_format.RunSpecs(&out, &out, []specs_format.RawSpec{
		{
			NumscriptPath:    "example.num",
			SpecsPath:        "example.num.specs.json",
			NumscriptContent: script,
			SpecsFileContent: []byte(specs),
		},
	})

	require.False(t, success)

	snaps.MatchSnapshot(t, out.String())
}

func TestComplexAssertions(t *testing.T) {
	var out bytes.Buffer

	script := `
		send [USD/2 100] (
			source = @alice
			destination = @dest
		)
	`

	specs := `
		{
			"testCases": [
				{
					"it": "send when there are enough funds",
					"balances": {
						"alice": { "USD/2": 9999 }
					},
					"expect.volumes": {
							"alice": { "USD/2": -100 },
							"dest": { "USD/2": 1 }
					},
					"expect.movements": {
						"alice": {
							"dest": { "EUR": 100 }
						}
					},
					"expect.missingFunds": true
				},
				{
					"it": "tpassing",
					"balances": {
						"alice": { "USD/2": 0 }
					},
					"expect.missingFunds": true
				}
			] 
		}
	`

	success := specs_format.RunSpecs(&out, &out, []specs_format.RawSpec{
		{
			NumscriptPath:    "example.num",
			SpecsPath:        "example.num.specs.json",
			NumscriptContent: script,
			SpecsFileContent: []byte(specs),
		},
	})

	require.False(t, success)

	snaps.MatchSnapshot(t, out.String())
}

func TestNoFilesErr(t *testing.T) {
	var out bytes.Buffer
	success := specs_format.RunSpecs(&out, &out, []specs_format.RawSpec{})
	require.False(t, success)
	snaps.MatchSnapshot(t, out.String())
}

func TestParseErrSpecs(t *testing.T) {
	var out bytes.Buffer

	success := specs_format.RunSpecs(&out, &out, []specs_format.RawSpec{
		{
			NumscriptPath:    "example.num",
			SpecsPath:        "example.num.specs.json",
			NumscriptContent: "",
			SpecsFileContent: []byte(`
		not a json
	`),
		},
	})
	require.False(t, success)
	snaps.MatchSnapshot(t, out.String())
}

func TestSchemaErrSpecs(t *testing.T) {
	var out bytes.Buffer

	success := specs_format.RunSpecs(&out, &out, []specs_format.RawSpec{
		{
			NumscriptPath:    "example.num",
			SpecsPath:        "example.num.specs.json",
			NumscriptContent: "",
			SpecsFileContent: []byte(`
		{ "balances": 42 }
	`),
		},
	})
	require.False(t, success)
	snaps.MatchSnapshot(t, out.String())
}

func TestNumscriptParseErr(t *testing.T) {
	var out bytes.Buffer

	success := specs_format.RunSpecs(&out, &out, []specs_format.RawSpec{
		{
			NumscriptPath:    "example.num",
			SpecsPath:        "example.num.specs.json",
			NumscriptContent: "!err",
			SpecsFileContent: []byte(`
		{ }
	`),
		},
	})
	require.False(t, success)
	snaps.MatchSnapshot(t, out.String())
}

func TestRuntimeErr(t *testing.T) {
	var out bytes.Buffer

	specs := `
		{
			"testCases": [
				{
					"it": "runs",
					"expect.missingFunds": false
				}
			] 
		}
	`

	success := specs_format.RunSpecs(&out, &out, []specs_format.RawSpec{
		{
			NumscriptPath:    "example.num",
			SpecsPath:        "example.num.specs.json",
			NumscriptContent: `send [USD/2 100] ( source = "ops!" destination = @world)`,
			SpecsFileContent: []byte(specs),
		},
	})
	require.False(t, success)
	snaps.MatchSnapshot(t, out.String())
}
