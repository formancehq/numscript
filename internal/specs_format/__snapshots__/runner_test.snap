
[TestShowDiff - 1]
[90m  {[0m
[90m    "common": "ok",[0m
[32m-   "x": 42[0m
[31m+   "x": 100[0m
[90m  }[0m

---

[TestSingleTest - 1]
[31m❯[0m example.num [90m(2 tests | [31m1 failed[0m[90m)[0m
  [31m×[0m tfailing


[1m[97m[41m FAIL [0m[97m[0m[1m[0m[31m example.num.specs.json > tfailing[0m
[4m
GOT:[0m

| [36mSource[0m | [36mDestination[0m | [36mAsset[0m | [36mAmount[0m |
| world  | dest        | USD/2 | 100    |


[32m- Expected[0m
[31m+ Received
[0m
[4mexpect.postings[0m

[90m  [[0m
[90m    {[0m
[32m-     "source": "wrong-source",[0m
[31m+     "source": "world",[0m
[90m      "destination": "dest",[0m
[90m      "amount": 100,[0m
[90m      "asset": "USD/2"[0m
[90m    }[0m
[90m  ][0m


[90m Test files [0m [1m[91m1 failed[0m[1m[0m [90m(1)[0m
[90m      Tests [0m [1m[91m1 failed[0m[1m[0m[90m | [0m[1m[92m1 passed[0m[1m[0m [90m(2)[0m

---

[TestComplexAssertions - 1]
[31m❯[0m example.num [90m(2 tests | [31m1 failed[0m[90m)[0m
  [31m×[0m send when there are enough funds


[1m[97m[41m FAIL [0m[97m[0m[1m[0m[31m example.num.specs.json > send when there are enough funds[0m
[4m
GIVEN:[0m

| [36mAccount[0m | [36mAsset[0m | [36mBalance[0m |
| alice   | USD/2 | 9999    |

[4m
GOT:[0m

| [36mSource[0m | [36mDestination[0m | [36mAsset[0m | [36mAmount[0m |
| alice  | dest        | USD/2 | 100    |


[32m- Expected[0m
[31m+ Received
[0m
[4mexpect.error.missingFunds[0m

[32m- true[0m
[31m+ false[0m

[4mexpect.endBalances[0m

[90m  {[0m
[90m    "alice": {[0m
[32m-     "USD/2": -100[0m
[31m+     "USD/2": 9899[0m
[90m    },[0m
[90m    "dest": {[0m
[32m-     "USD/2": 1[0m
[31m+     "USD/2": 100[0m
[90m    }[0m
[90m  }[0m

[4mexpect.movements[0m

[90m  {[0m
[90m    "alice": {[0m
[90m      "dest": {[0m
[32m-       "EUR": 100[0m
[31m+       "USD/2": 100[0m
[90m      }[0m
[90m    }[0m
[90m  }[0m


[90m Test files [0m [1m[91m1 failed[0m[1m[0m [90m(1)[0m
[90m      Tests [0m [1m[91m1 failed[0m[1m[0m[90m | [0m[1m[92m1 passed[0m[1m[0m [90m(2)[0m

---

[TestNoFilesErr - 1]
[31mNo specs files found
[0m
---

[TestParseErrSpecs - 1]
[31m
Error: example.num.specs.json

[0minvalid character 'o' in literal null (expecting 'u')

---

[TestSchemaErrSpecs - 1]
[31m
Error: example.num.specs.json

[0mjson: cannot unmarshal number into Go struct field Specs.balances of type interpreter.Balances

---

[TestNumscriptParseErr - 1]
[31m
Error: example.num:1:1

[0mtoken recognition error at: '!'

[31m
Error: example.num:1:5

[0mmismatched input '<EOF>' expecting '('


  0 | !err
    |     ~~~~

---

[TestRuntimeErr - 1]
[31m
Error: example.num:1:29

[0mInvalid value received. Expecting value of type account (got ops! instead)


  0 | send [USD/2 100] ( source = "ops!" destination = @world)
    |                             ~~~~~~

---
